src/packages/auth/functions.ts

```diff
@@ -46,28 +55,53 @@ export async function getUserRoles(
     const { hasura_allowed_roles, hasura_default_role } = row;
     return { allowed_roles: hasura_allowed_roles, default_role: hasura_default_role };
   } else {
+    // since user does not exist, this upsert is just an insert
+    upsertUserRoles(username, default_role, allowed_roles);
+    return { allowed_roles, default_role };
+  }
+}
+
+export async function deleteUserAllowedRoles(username: string) {
+  const db = DbMerlin.getDb();
+
+  await db.query(
+    `
+      delete from metadata.users_allowed_roles
+      where username = $1;
+    `,
+    [username],
+  );
+}
+
+export async function upsertUserRoles(username: string, default_role: string, allowed_roles: string[]) {
+  const db = DbMerlin.getDb();
+
+  await db.query(
+    `
+      insert into metadata.users (username, default_role)
+      values ($1, $2)
+      on conflict (username) do update
+      set default_role = excluded.default_role;
+    `,
+    [username, default_role],
+  );
+
+  for (const allowed_role of allowed_roles) {
     await db.query(
       `
-        insert into metadata.users (username, default_role)
-        values ($1, $2);
+        insert into metadata.users_allowed_roles (username, allowed_role)
+        values ($1, $2)
       `,
-      [username, default_role],
+      [username, allowed_role],
     );
-
-    for (const allowed_role of allowed_roles) {
-      await db.query(
-        `
-          insert into metadata.users_allowed_roles (username, allowed_role)
-          values ($1, $2);
-        `,
-        [username, allowed_role],
-      );
-    }
-
-    return { allowed_roles, default_role };
   }
 }
 
+export async function syncRolesToDB(username: string, default_role: string, allowed_roles: string[]) {
+  await deleteUserAllowedRoles(username);
+  await upsertUserRoles(username, default_role, allowed_roles);
+}
+
```

Comment:
Nonblocking but would _strongly_ prefer if these were in a single transaction instead of 2.

Reply:
Agreed, they originally were, but I removed them while debugging and never restored them

Reply:
Change has been made.